File: motor/src/protocolo.cpp
/*  Pletina DAW · motor
    Copyright (C) 2026 Up2you — GPLv3: ver el LICENSE de la raíz.

    Traducción entre líneas NDJSON y órdenes del Motor. El formato está descrito
    en protocolo.h y en docs/04-protocolo.md.
*/

#include "protocolo.h"
#include "motor.h"

namespace
{
    juce::var objeto()
    {
        return juce::var (new juce::DynamicObject());
    }

    void pon (juce::var& v, const juce::Identifier& clave, const juce::var& valor)
    {
        v.getDynamicObject()->setProperty (clave, valor);
    }

    juce::String aLinea (const juce::var& v)
    {
        return juce::JSON::toString (v, true);
    }
}

namespace protocolo
{
    juce::String evento (const juce::String& nombre, const juce::var& datos)
    {
        auto v = objeto();
        pon (v, "evento", nombre);
        pon (v, "datos", datos);
        return aLinea (v);
    }

    juce::String respuesta (const juce::var& id, const juce::var& resultado)
    {
        auto v = objeto();
        pon (v, "id", id);
        pon (v, "resultado", resultado);
        return aLinea (v);
    }

    juce::String error (const juce::var& id, int codigo, const juce::String& mensaje)
    {
        auto e = objeto();
        pon (e, "codigo", codigo);
        pon (e, "mensaje", mensaje);

        auto v = objeto();
        pon (v, "id", id);
        pon (v, "error", e);
        return aLinea (v);
    }

    juce::String procesarLinea (Motor& motor, const juce::String& linea, bool& pidieronSalir)
    {
        if (linea.trim().isEmpty())
            return {};

        auto mensaje = juce::JSON::parse (linea);

        if (! mensaje.isObject())
            return error (juce::var(), -32700, "no es JSON: " + linea.substring (0, 80));

        const auto id = mensaje["id"];
        const auto metodo = mensaje["metodo"].toString();
        const auto params = mensaje["params"];

        try
        {
            if (metodo == "hola")                 return respuesta (id, motor.hola());
            if (metodo == "dispositivos.listar")  return respuesta (id, motor.listarDispositivos());

            if (metodo == "proyecto.nuevo")       return respuesta (id, motor.nuevoProyecto (params["carpeta"].toString()));
            if (metodo == "proyecto.abrir")       return respuesta (id, motor.abrirProyecto (params["ruta"].toString()));
            if (metodo == "proyecto.guardar")     return respuesta (id, motor.guardarProyecto());
            if (metodo == "pistas.listar")        return respuesta (id, motor.listarPistas());

            if (metodo == "pista.crear")          return respuesta (id, motor.crearPista());
            if (metodo == "pista.borrar")         return respuesta (id, motor.borrarPista ((int) params["pista"]));
            if (metodo == "pista.renombrar")      return respuesta (id, motor.renombrarPista ((int) params["pista"], params["nombre"].toString()));
            if (metodo == "pista.mezcla")         return respuesta (id, motor.mezclaPista (params.hasProperty ("pista") ? (int) params["pista"] : -1, params));
            if (metodo == "pista.envio")          return respuesta (id, motor.envioPista ((int) params["pista"], (int) params["bus"], (double) params["nivelDb"]));
            if (metodo == "pista.congelar")       return respuesta (id, motor.congelarPista ((int) params["pista"], (bool) params["activo"]));

            if (metodo == "clip.importar")        return respuesta (id, motor.importarClip ((int) params["pista"], params["ruta"].toString(), (double) params["inicio"]));
            if (metodo == "clip.mover")           return respuesta (id, motor.moverClip (params["id"].toString(), (double) params["inicio"], params.hasProperty ("pista") ? (int) params["pista"] : -1));
            if (metodo == "clip.recortar")        return respuesta (id, motor.recortarClip (params["id"].toString(), (double) params["inicio"], (double) params["fin"]));
            if (metodo == "clip.dividir")         return respuesta (id, motor.dividirClip (params["id"].toString(), (double) params["segundos"]));
            if (metodo == "clip.duplicar")        return respuesta (id, motor.duplicarClip (params["id"].toString()));
            if (metodo == "clip.fundidos")        return respuesta (id, motor.fundidosClip (params["id"].toString(), (double) params["entrada"], (double) params["salida"]));
            if (metodo == "clip.borrar")          return respuesta (id, motor.borrarClip (params["id"].toString()));
            if (metodo == "clip.picos")           return respuesta (id, motor.picosClip (params["id"].toString(), params.hasProperty ("porSegundo") ? (int) params["porSegundo"] : 50));
            if (metodo == "clip.warp")            return respuesta (id, motor.warpClip (params["id"].toString(), params));

            if (metodo == "dispositivos.tono")    return respuesta (id, motor.tonoDePrueba (params));
            if (metodo == "vst.carpetas")         return respuesta (id, motor.carpetasVst (params));
            if (metodo == "vst.escanear")         return respuesta (id, motor.escanearVst());
            if (metodo == "vst.lista")            return respuesta (id, motor.listaVst());
            if (metodo == "previa.tocar")         return respuesta (id, motor.tocarPrevia (params["ruta"].toString()));
            if (metodo == "previa.parar")         return respuesta (id, motor.pararPrevia());
            if (metodo == "pista.armar")          return respuesta (id, motor.armarPista ((int) params["pista"], (bool) params["activo"], params.hasProperty ("entrada") ? (int) params["entrada"] : 0, params.hasProperty ("midi") && (bool) params["midi"]));
            if (metodo == "transporte.grabar")    return respuesta (id, motor.grabar (params));

            if (metodo == "grupo.crear")          return respuesta (id, motor.crearGrupo (params));
            if (metodo == "grupo.meter")          return respuesta (id, motor.meterEnGrupo ((int) params["pista"], (int) params["grupo"]));
            if (metodo == "grupo.sacar")          return respuesta (id, motor.sacarDeGrupo ((int) params["pista"]));
            if (metodo == "grupo.deshacer")       return respuesta (id, motor.deshacerGrupo ((int) params["grupo"]));

            if (metodo == "clip.midi.crear")      return respuesta (id, motor.crearClipMidi ((int) params["pista"], (double) params["inicio"], params.hasProperty ("compases") ? (double) params["compases"] : 1.0));
            if (metodo == "clip.midi.notas")      return respuesta (id, motor.notasClipMidi (params["id"].toString(), params["notas"]));
            if (metodo == "clip.midi.cuantizar")  return respuesta (id, motor.cuantizarClipMidi (params["id"].toString(), params["division"].toString()));

            if (metodo == "sesion.escenas")       return respuesta (id, motor.escenasSesion ((int) params["numero"]));
            if (metodo == "sesion.poner")         return respuesta (id, motor.ponerEnSesion ((int) params["pista"], (int) params["escena"], params["desdeClip"].toString()));
            if (metodo == "sesion.lanzar")        return respuesta (id, motor.lanzarSesion (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["escena"]));
            if (metodo == "sesion.parar")         return respuesta (id, motor.pararSesion (params.hasProperty ("pista") ? (int) params["pista"] : -1));
            if (metodo == "sesion.cuantizacion")  return respuesta (id, motor.cuantizacionSesion (params["nombre"].toString()));

            if (metodo == "plugin.insertar")      return respuesta (id, motor.insertarPlugin (params.hasProperty ("pista") ? (int) params["pista"] : -1, params["tipo"].toString(), params.hasProperty ("indice") ? (int) params["indice"] : -1));
            if (metodo == "plugin.quitar")        return respuesta (id, motor.quitarPlugin (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"]));
            if (metodo == "plugin.parametro")     return respuesta (id, motor.parametroPlugin (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], params["parametro"].toString(), (double) params["valor"]));
            if (metodo == "plugin.activar")       return respuesta (id, motor.activarPlugin (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], (bool) params["activo"]));
            if (metodo == "plugin.lateral")       return respuesta (id, motor.lateralPlugin (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], params.hasProperty ("fuente") ? (int) params["fuente"] : -1));
            if (metodo == "plugin.presets")       return respuesta (id, motor.listarPresets (params["tipo"].toString()));
            if (metodo == "plugin.preset.guardar") return respuesta (id, motor.guardarPreset (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], params["nombre"].toString()));
            if (metodo == "plugin.preset.cargar")  return respuesta (id, motor.cargarPreset (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], params["nombre"].toString()));

            if (metodo == "rack.crear")           return respuesta (id, motor.crearRack (params.hasProperty ("pista") ? (int) params["pista"] : -1, params.hasProperty ("desde") ? (int) params["desde"] : -1, params.hasProperty ("hasta") ? (int) params["hasta"] : -1, params["nombre"].toString()));
            if (metodo == "rack.deshacer")        return respuesta (id, motor.deshacerRack (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"]));
            if (metodo == "rack.macro")           return respuesta (id, motor.macroRack (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], (int) params["macro"], params));
            if (metodo == "rack.asignar")         return respuesta (id, motor.asignarMacroRack (params.hasProperty ("pista") ? (int) params["pista"] : -1, (int) params["indice"], (int) params["macro"], params));

            if (metodo == "automatizacion.puntos") return respuesta (id, motor.puntosAutomatizacion (params));

            if (metodo == "transporte.tocar")     return respuesta (id, motor.tocar());
            if (metodo == "transporte.parar")     return respuesta (id, motor.parar());
            if (metodo == "transporte.irA")       return respuesta (id, motor.irA ((double) params["segundos"]));
            if (metodo == "transporte.estado")    return respuesta (id, motor.estadoTransporte());
            if (metodo == "transporte.tempo")     return respuesta (id, motor.tempo ((double) params["bpm"]));
            if (metodo == "transporte.metronomo") return respuesta (id, motor.metronomo ((bool) params["activo"]));
            if (metodo == "transporte.bucle")     return respuesta (id, motor.bucle ((bool) params["activo"], (double) params["inicio"], (double) params["fin"]));

            if (metodo == "deshacer.deshacer")    return respuesta (id, motor.deshacer());
            if (metodo == "deshacer.rehacer")     return respuesta (id, motor.rehacer());

            if (metodo == "render.exportar")      return respuesta (id, motor.exportar (params["ruta"].toString(),
                                                                     params.hasProperty ("stems") && (bool) params["stems"],
                                                                     params.hasProperty ("lufsObjetivo") ? (double) params["lufsObjetivo"] : -1000.0));

            if (metodo == "salir")
            {
                pidieronSalir = true;
                auto r = objeto();
                pon (r, "adios", true);
                return respuesta (id, r);
            }
        }
        catch (const std::exception& e)
        {
            return error (id, -32000, e.what());
        }

        return error (id, -32601, "método desconocido: " + metodo);
    }
}

File: ui/src/shared/protocolo.js
/**
 * El protocolo UI⇄motor, visto desde este lado: líneas JSON (NDJSON) por el
 * stdio del proceso `pletina-motor`. El formato completo está en
 * docs/04-protocolo.md.
 *
 * La lista de métodos vive dos veces —aquí y en motor/src/protocolo.h— y el
 * test de contrato (test/contrato.test.js) comprueba que son idénticas:
 * añadir un método en un solo lado rompe la build, a propósito.
 */

export const METODOS = [
  'hola',
  'dispositivos.listar',
  'dispositivos.tono',
  'vst.carpetas',
  'vst.escanear',
  'vst.lista',
  'previa.tocar',
  'previa.parar',
  'proyecto.nuevo',
  'proyecto.abrir',
  'proyecto.guardar',
  'pistas.listar',
  'pista.crear',
  'pista.borrar',
  'pista.renombrar',
  'pista.mezcla',
  'pista.envio',
  'pista.congelar',
  'pista.armar',
  'clip.importar',
  'clip.mover',
  'clip.recortar',
  'clip.dividir',
  'clip.duplicar',
  'clip.fundidos',
  'clip.borrar',
  'clip.picos',
  'clip.warp',
  'clip.midi.crear',
  'clip.midi.notas',
  'clip.midi.cuantizar',
  'sesion.escenas',
  'sesion.poner',
  'sesion.lanzar',
  'sesion.parar',
  'sesion.cuantizacion',
  'plugin.insertar',
  'plugin.quitar',
  'plugin.parametro',
  'plugin.activar',
  'plugin.lateral',
  'plugin.presets',
  'plugin.preset.guardar',
  'plugin.preset.cargar',
  'automatizacion.puntos',
  'transporte.tocar',
  'transporte.grabar',
  'transporte.parar',
  'transporte.irA',
  'transporte.estado',
  'transporte.tempo',
  'transporte.metronomo',
  'transporte.bucle',
  'deshacer.deshacer',
  'deshacer.rehacer',
  'render.exportar',
  'salir',
];

export const EVENTOS = ['arrancado', 'medidores', 'modelo', 'render.terminado', 'prueba'];

/** Serializa una petición a línea NDJSON. El id lo pone quien lleva la cuenta. */
export function codificarPeticion(id, metodo, params) {
  if (!METODOS.includes(metodo)) throw new Error(`método fuera del protocolo: ${metodo}`);
  const mensaje = params === undefined ? { id, metodo } : { id, metodo, params };
  return `${JSON.stringify(mensaje)}\n`;
}

/**
 * Clasifica una línea recibida del motor. Devuelve `{tipo, ...}` con tipo
 * `respuesta` (con `id` y `resultado` o `error`), `evento` (con `nombre` y
 * `datos`) o `ruido` (una línea que no es del protocolo: se ignora avisando).
 */
export function decodificarLinea(linea) {
  const texto = String(linea).trim();
  if (!texto) return { tipo: 'ruido', linea: texto };

  let mensaje;
  try {
    mensaje = JSON.parse(texto);
  } catch {
    return { tipo: 'ruido', linea: texto };
  }

  if (mensaje && typeof mensaje === 'object' && 'evento' in mensaje) {
    return { tipo: 'evento', nombre: mensaje.evento, datos: mensaje.datos ?? {} };
  }
  if (mensaje && typeof mensaje === 'object' && 'id' in mensaje) {
    return { tipo: 'respuesta', id: mensaje.id, resultado: mensaje.resultado, error: mensaje.error };
  }
  return { tipo: 'ruido', linea: texto };
}

/**
 * Trocea el flujo de stdout en líneas completas. Devuelve las líneas cerradas
 * y el resto pendiente, porque un `data` puede cortar una línea por la mitad.
 */
export function partirLineas(pendiente, trozo) {
  const texto = pendiente + trozo;
  const partes = texto.split('\n');
  const resto = partes.pop();
  return { lineas: partes.filter((l) => l.trim() !== ''), resto };
}
